### PR TITLE
body recursive fetch_all without Enum.reverse

### LINKED ISSUE #204: fetch_all breaks row order
👋

When using `ecto_sqlite3` with a table and an index

```sql
CREATE TABLE IF NOT EXISTS "heartbeats" ("time" REAL NOT NULL, "entity" TEXT NOT NULL, "type" TEXT NOT NULL, "category" TEXT, "project" TEXT, "branch" TEXT, "language" TEXT, "dependencies" TEXT, "lines" INTEGER, "lineno" INTEGER, "cursorpos" INTEGER, "is_write" INTEGER DEFAULT false NOT NULL, "editor" TEXT, "operating_system" TEXT) strict;
CREATE INDEX "heartbeats_time_index" ON "heartbeats" ("time");
```

executing 

```sql
SELECT time, project FROM heartbeats WHERE (time > ?) AND (time < ?) ORDER BY time [1654026000, 1654074000]
```

returns rows unordered by time from `fetch_all`. The rows are ordered by time when iterating `step` and `multi_step` manually.

### Steps to reproduce:

[w2_dev.csv](https://github.com/elixir-sqlite/exqlite/files/8812687/w2_dev.csv)

```sh
> curl https://github.com/elixir-sqlite/exqlite/files/8812687/w2_dev.csv -o w2_dev.csv
> sqlite3 fetch_all_issue.db
```

```sql
sqlite> CREATE TABLE IF NOT EXISTS "heartbeats" ("time" REAL NOT NULL, "entity" TEXT NOT NULL, "type" TEXT NOT NULL, "category" TEXT, "project" TEXT, "branch" TEXT, "language" TEXT, "dependencies" TEXT, "lines" INTEGER, "lineno" INTEGER, "cursorpos" INTEGER, "is_write" INTEGER DEFAULT false NOT NULL, "editor" TEXT, "operating_system" TEXT) strict;
sqlite> CREATE INDEX "heartbeats_time_index" ON "heartbeats" ("time");
sqlite> .import w2_dev.csv heartbeats --csv
sqlite> select count(*) from heartbeats;
925
```

```sh
> iex
```

```elixir
iex> Mix.install([:exqlite])
iex> alias Exqlite.Sqlite3
iex> {:ok, conn} = Sqlite3.open("fetch_all_issue.db")
iex> {:ok, stmt} = Sqlite3.prepare(conn, "SELECT time, project FROM heartbeats WHERE (time > ?) AND (time < ?) ORDER BY time")
iex> :ok = Sqlite3.bind(conn, stmt, [1654026000, 1654074000])
iex> {:ok, rows} = Sqlite3.fetch_all(conn, stmt)
iex> rows |> Enum.map(fn [time, _] -> DateTime.from_unix!(floor(time)) end) |> IO.inspect(limit: :infinity)

[~U[2022-05-31 19:40:07Z], ~U[2022-05-31 19:40:08Z], ~U[2022-05-31 19:40:09Z],
 ~U[2022-05-31 19:40:09Z], ~U[2022-05-31 19:40:09Z], ~U[2022-05-31 19:40:10Z],
 ~U[2022-05-31 19:40:10Z], ~U[2022-05-31 19:40:10Z], ~U[2022-05-31 19:40:11Z],
 ~U[2022-05-31 19:40:11Z], ~U[2022-05-31 19:40:12Z], ~U[2022-05-31 19:40:17Z],
 ~U[2022-05-31 19:40:21Z], ~U[2022-05-31 19:40:36Z], ~U[2022-05-31 19:40:59Z],
 ~U[2022-05-31 19:41:17Z], ~U[2022-05-31 19:41:54Z], ~U[2022-05-31 19:42:02Z],
 ~U[2022-05-31 19:42:17Z], ~U[2022-05-31 19:42:52Z], ~U[2022-05-31 19:43:24Z],
 ~U[2022-05-31 19:43:27Z], ~U[2022-05-31 19:43:40Z], ~U[2022-05-31 19:43:47Z],
 ~U[2022-05-31 19:44:54Z], ~U[2022-05-31 19:45:32Z], ~U[2022-05-31 19:47:32Z],
 ~U[2022-05-31 19:49:33Z], ~U[2022-05-31 19:50:47Z], ~U[2022-05-31 19:50:59Z],
 ~U[2022-05-31 19:51:08Z], ~U[2022-05-31 19:51:15Z], ~U[2022-05-31 19:51:33Z],
 ~U[2022-05-31 19:52:15Z], ~U[2022-05-31 19:52:19Z], ~U[2022-05-31 19:52:22Z],
 ~U[2022-05-31 19:52:25Z], ~U[2022-05-31 19:52:27Z], ~U[2022-05-31 19:52:59Z],
 ~U[2022-05-31 19:55:02Z], ~U[2022-05-31 19:55:11Z], ~U[2022-05-31 19:56:34Z],
 ~U[2022-05-31 19:56:50Z], ~U[2022-05-31 19:57:12Z], ~U[2022-05-31 19:57:43Z],
 ~U[2022-05-31 19:58:54Z], ~U[2022-05-31 20:01:11Z], ~U[2022-05-31 20:03:35Z],
 ~U[2022-05-31 20:05:27Z], ~U[2022-05-31 20:07:30Z], ~U[2022-05-31 20:08:36Z],
 ~U[2022-05-31 20:10:02Z], ~U[2022-05-31 20:10:03Z], ~U[2022-05-31 20:11:11Z],
 ~U[2022-05-31 20:11:35Z], ~U[2022-05-31 20:12:12Z], ~U[2022-05-31 20:13:49Z],
 ~U[2022-05-31 20:15:42Z], ~U[2022-05-31 20:16:00Z], ~U[2022-05-31 20:16:40Z],
 ~U[2022-05-31 20:16:44Z], ~U[2022-05-31 20:18:03Z], ~U[2022-05-31 20:18:51Z],
 ~U[2022-05-31 20:19:42Z], ~U[2022-05-31 20:20:12Z], ~U[2022-05-31 20:20:32Z],
 ~U[2022-05-31 20:20:37Z], ~U[2022-05-31 20:22:47Z], ~U[2022-06-01 08:10:03Z],
 ~U[2022-06-01 08:10:05Z], ~U[2022-06-01 08:12:26Z], ~U[2022-06-01 08:14:45Z],
 ~U[2022-06-01 08:15:33Z], ~U[2022-06-01 08:17:26Z], ~U[2022-06-01 08:18:37Z],
 ~U[2022-06-01 08:18:41Z], ~U[2022-06-01 08:20:44Z], ~U[2022-06-01 08:21:17Z],
 ~U[2022-06-01 08:26:44Z], ~U[2022-06-01 08:28:18Z], ~U[2022-06-01 08:29:35Z],
 ~U[2022-06-01 08:29:51Z], ~U[2022-06-01 08:29:57Z], ~U[2022-06-01 08:30:00Z],
 ~U[2022-06-01 08:30:03Z], ~U[2022-06-01 08:30:07Z], ~U[2022-06-01 08:30:10Z],
 ~U[2022-06-01 08:30:26Z], ~U[2022-06-01 08:30:45Z], ~U[2022-06-01 08:30:54Z],
 ~U[2022-06-01 08:33:32Z], ~U[2022-06-01 08:33:34Z], ~U[2022-06-01 08:34:11Z],
 ~U[2022-06-01 08:36:02Z], ~U[2022-06-01 08:36:43Z], ~U[2022-06-01 08:37:19Z],
 ~U[2022-06-01 08:37:27Z], ~U[2022-06-01 08:37:50Z], ~U[2022-06-01 08:38:07Z],
 ~U[2022-06-01 08:39:51Z], ~U[2022-06-01 08:53:34Z], ~U[2022-06-01 08:53:32Z], # order breaks here
 ~U[2022-06-01 08:52:59Z], ~U[2022-06-01 08:51:18Z], ~U[2022-06-01 08:51:01Z],
 ~U[2022-06-01 08:50:24Z], ~U[2022-06-01 08:49:56Z], ~U[2022-06-01 08:49:50Z],
 ~U[2022-06-01 08:44:05Z], ~U[2022-06-01 08:43:50Z], ~U[2022-06-01 08:43:49Z],
 ~U[2022-06-01 08:43:46Z], ~U[2022-06-01 08:43:40Z], ~U[2022-06-01 08:43:31Z],
 ~U[2022-06-01 08:43:09Z], ~U[2022-06-01 08:43:03Z], ~U[2022-06-01 08:43:00Z],
 ~U[2022-06-01 08:42:46Z], ~U[2022-06-01 08:42:06Z], ~U[2022-06-01 08:41:38Z],
 ~U[2022-06-01 08:41:34Z], ~U[2022-06-01 08:41:06Z], ~U[2022-06-01 08:40:51Z],
 ~U[2022-06-01 08:40:48Z], ~U[2022-06-01 08:40:33Z], ~U[2022-06-01 08:40:10Z]]
```

### PR DESCRIPTION
**Pull request by ruslandoga**
closes #204

### CODE REVIEW

File: lib/exqlite/sqlite3.ex [hunk 1]
```diff
@@ -129,26 +140,7 @@ defmodule Exqlite.Sqlite3 do
...
     chunk_size = Application.get_env(:exqlite, :default_chunk_size, 50)
-    fetch_all(conn, statement, chunk_size, [])
-  end
-
-  defp fetch_all(conn, statement, chunk_size, accum) do
-    case multi_step(conn, statement, chunk_size) do
-      {:done, rows} ->
-        case accum do
-          [] -> {:ok, rows}
-          accum -> {:ok, Enum.reverse(rows ++ accum)}
```
Comment (ruslandoga) [inline lib/exqlite/sqlite3.ex:140]: rows should've been reversed here before concating, I think

**Comment by ruslandoga** [comment]
`bench.exs`

```elixir
Mix.install([:benchee, :exqlite])

alias Exqlite.Sqlite3
{:ok, conn} = Sqlite3.open(":memory:")
:ok = Sqlite3.execute(conn, "create table users (id integer primary key, name text)")

{:ok, stmt} = Sqlite3.prepare(conn, "insert into users values (?, ?)")

Sqlite3.execute(conn, "begin")

Enum.each(1..1_000, fn i ->
  :ok = Sqlite3.bind(conn, stmt, [i, "User-" <> to_string(i)])
  :done = Sqlite3.step(conn, stmt)
end)

Sqlite3.execute(conn, "commit")

defmodule Tail do
  def fetch_all(conn, statement) do
    fetch_all(conn, statement, 50, [])
  end

  defp fetch_all(conn, statement, chunk_size, accum) do
    case Sqlite3.multi_step(conn, statement, chunk_size) do
      {:done, rows} ->
        case accum do
          [] -> {:ok, rows}
          accum -> {:ok, Enum.reverse(rows ++ accum)}
        end

      {:rows, rows} ->
        fetch_all(conn, statement, chunk_size, Enum.reverse(rows) ++ accum)

      {:error, reason} ->
        {:error, reason}

      :busy ->
        {:error, "Database busy"}
    end
  end
end

defmodule Body do
  def fetch_all(conn, statement) do
    {:ok, try_fetch_all(conn, statement, 50)}
  catch
    :throw, {:error, _reason} = error -> error
  end

  defp try_fetch_all(conn, statement, chunk_size) do
    case Sqlite3.multi_step(conn, statement, chunk_size) do
      {:done, rows} -> rows
      {:rows, rows} -> rows ++ try_fetch_all(conn, statement, chunk_size)
      {:error, _reason} = error -> throw(error)
      :busy -> throw({:error, "Database busy"})
    end
  end
end

{:ok, stmt} = Sqlite3.prepare(conn, "select * from users")

Benchee.run(
  %{
    "body-recursive" => fn -> Body.fetch_all(conn, stmt) end,
    "tail-recursive" => fn -> Tail.fetch_all(conn, stmt) end
  },
  memory_time: 2
)
```

```
> elixir bench.exs
```

File: lib/exqlite/sqlite3.ex (hunk 1 above)
Comment (warmwaffles) [reply lib/exqlite/sqlite3.ex:140]: You are right. They should have been reversed, before appending.

**Comment by SebastianSzturo** [comment]
Thanks for fixing this properly! Looks like the initial bottle neck was not the body recursion 😅

**Pull request merged**